File: elixir/tournament/lib/tournament.ex
defmodule Tournament do
  @header "Team                           | MP |  W |  D |  L |  P"

  defmodule Result do
    defstruct [team: "", matches_played: 0, wins: 0, draws: 0, losses: 0, points: 0]

    def empty, do: %Result{}
    def win, do: %Result{matches_played: 1, wins: 1, draws: 0, losses: 0, points: 3}
    def loss, do: %Result{matches_played: 1, wins: 0, draws: 0, losses: 1, points: 0}
    def draw, do: %Result{matches_played: 1, wins: 0, draws: 1, losses: 0, points: 1}
  end

  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """
  @spec tally(input :: list(String.t())) :: String.t()
  def tally([]), do: @header
  def tally(input) when is_binary(input) do
    try do
      [t1, t2, result] = String.split(input, ";", trim: true)
      format_result_lines(t1, t2, result)
    rescue 
      MatchError -> []
    end
  end

  def tally(input) do
    lines = input
    |> Enum.map(&tally/1)
    |> List.flatten()
    |> Enum.group_by(&(&1.team))
    |> Map.values()
    |> Enum.map(&merge_results/1)
    |> Enum.sort(&(&1.points >= &2.points))
    |> Enum.map(&format_line/1)
    |> Enum.join("\n")

    "#{@header}\n#{lines}"
  end

  defp merge_result_fields(:__struct__, _, _), do: nil
  defp merge_result_fields(:team, _, value), do: value
  defp merge_result_fields(_, value1, value2), do: value1 + value2

  defp merge_result_records(result1, result2) do
    Map.merge(result1, result2, &merge_result_fields/3)
  end

  defp merge_results(results) do
    Enum.reduce(results, Result.empty, fn result, consolidation -> merge_result_records(consolidation, result) end)
  end

  defp format_result_lines(t1, t2, "win") do
    [%Result{Result.win() | team: t1}, %Result{Result.loss() | team: t2}]
  end

  defp format_result_lines(t1, t2, "loss") do
    [%Result{Result.win() | team: t2}, %Result{Result.loss() | team: t1}]
  end

  defp format_result_lines(t1, t2, "draw") do
    [%Result{Result.draw() | team: t2}, %Result{Result.draw() | team: t1}]
  end

  defp format_result_lines(_, _, _), do: []

  defp format_line(results) do
    "#{String.pad_trailing(results.team, 31)}|  #{results.matches_played} |  #{results.wins} |  #{results.draws} |  #{results.losses} |#{String.pad_leading(to_string(results.points), 3)}"
  end
end

File: solutions/ruby/tournament/1/tournament.rb
module Tournament
    class TeamResults
        def initialize(team)
            @team = team
            @played = 0
            @wins = 0
            @draws = 0
            @losses = 0
            @points = 0
        end

        def record_win()
            @played += 1
            @wins += 1
            @points += 3

            self
        end

        def record_loss()
            @played += 1
            @losses += 1

            self
        end

        def record_draw()
            @played += 1
            @draws += 1
            @points += 1

            self
        end
        
        def to_s()
            "| #{played.to_s.rjust(2)} | #{wins.to_s.rjust(2)} | #{draws.to_s.rjust(2)} | #{losses.to_s.rjust(2)} | #{points.to_s.rjust(2)}"
        end

        attr_reader :wins, :losses, :points, :draws, :played, :team
    end

    def self.tally(input)

        teams = {}
        input.split("\n").each do 
            | line |
            team1, team2, result = line.split(";")
            
            case result
            when "win"
                team1_results = teams.fetch(team1, TeamResults.new(team1))
                teams[team1] = team1_results.record_win
                team2_results = teams.fetch(team2, TeamResults.new(team2))
                teams[team2] = team2_results.record_loss
            when "loss"
                team1_results = teams.fetch(team1, TeamResults.new(team1))
                teams[team1] = team1_results.record_loss
                team2_results = teams.fetch(team2, TeamResults.new(team2))
                teams[team2] = team2_results.record_win
            when "draw"
                team1_results = teams.fetch(team1, TeamResults.new(team1))
                teams[team1] = team1_results.record_draw
                team2_results = teams.fetch(team2, TeamResults.new(team2))
                teams[team2] = team2_results.record_draw
            end
        end

        table_order = teams.sort_by { | _team,  result | [-result.points, result.team] }
        records = table_order.map { | team, results | "#{team.ljust(30)} #{results.to_s}\n" }.join()
        
        "Team                           | MP |  W |  D |  L |  P\n" + records
    end
end